node: podresources: add rate limiting

ffromani opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
Implement server-side rate limit support for podresources API endpoint (which is served locally on the node through unix domain socket). This is a GA graduation blocker: https://github.com/kubernetes/enhancements/pull/3863


#### Which issue(s) this PR fixes:
related (but not sufficient to close) to[ k/e 3743](https://github.com/kubernetes/enhancements/issues/3743)

#### Special notes for your reviewer:

```release-note
add options to set rate limiting for the node-local kubelet podresource API
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- KEP: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/606-compute-device-assignment/README.md
```

ffromani committed:
api: add ratelimitconfiguration for kubelet endpoints

The podresources API is a node-local gRPC API exposed by the kubelet
using a UNIX-domain socket which allows client to query about compute
resources exclusively allocated to pods (devices, cpus...)

As part as the feature GA graduation, we identified the
requirement to add rate limiting to prevent DOS from buggy or malicious
clients [1][2].

So this change extends the KubeletConfiguration to allow to
configure the ratelimit parameters.

The interface intentionally mimics the parameters of the
golang/x/time/rate package [3], because it's simple and already being
used in the codebase.

Because of this, there is an interdependency between the rate limiter
configuration parameters. This is the reason why the rate limiting is
optional, with defaults to "no limits" for backward compatibility, but
if specified, all the rate limit configuration values must be given
(e.g. burst doesn't make much sense without frequency, see [3]).

+++

[1] https://github.com/kubernetes/enhancements/pull/3791
[2] https://github.com/kubernetes/enhancements/pull/3863
[3] https://pkg.go.dev/golang.org/x/time/rate#Limiter

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
api: kubelet codegen: re-run generators

Update code generation after adding RateLimitConfiguration,
using
```
./hack/update-codegen.sh
```

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani commented:
/retitle WIP: node: podresources: add rate limiting
/sig node

ffromani commented:
/retest

```
 === Failed
=== FAIL: vendor/k8s.io/apiserver/pkg/storage/storagebackend/factory TestCreateReadycheck/timeouts_if_response_time_higher_than_custom_timeout (1.00s)
    factory_test.go:242: healthcheck() missmatch want context deadline exceeded got etcd client connection not yet established
    --- FAIL: TestCreateReadycheck/timeouts_if_response_time_higher_than_custom_timeout (1.00s)
=== FAIL: vendor/k8s.io/apiserver/pkg/storage/storagebackend/factory TestCreateReadycheck (13.95s) 
```

moshe010 commented on pkg/kubelet/server/server.go:
```diff
@@ -209,12 +212,27 @@ func ListenAndServeKubeletReadOnlyServer(
+	CpusProvider      podresources.CPUsProvider
+	MemoryProvider    podresources.MemoryProvider
+}
+
```
maybe you should have 2 structs
1. PodResourcesConfig (MaxCallsPerSecond , MaxCallsBurst, Socket)
2. PodResourcesProviders (PodsProvider, DevicesProvider   ...)

The PodResourcesProviders provders can be pass to the podresources.NewV1PodResourcesServer and podresources.NewV1alpha1PodResourcesServer so it will look like  podresources.NewV1PodResourcesServer(providers)

ffromani replied:
I don't have strong feelings here and I can explore this direction

ffromani committed:
kubelet: podresources: pack parameters in a struct

To enable rate limiting, needed for GA graduation,
we need to pass more parameters to the already crowded
`ListenAndServePodresources` function.

To tidy up a bit, pack the parameters in a helper struct,
with no intended changes in behavior.

Signed-off-by: Francesco Romani <fromani@redhat.com>

ffromani committed:
kubelet: podresources: implement ratelimit

Implement the node-local podresources API rate limiting wiring
in the golang.org/x/time/rate#Limiter struct, which was already
used in the codebase.

Add e2e test to exercise the flow, because the wiring itself
is mostly boilerplate and API adaptation.

ffromani commented on pkg/kubelet/kubelet.go:
```diff
@@ -2522,7 +2522,15 @@ func (kl *Kubelet) ListenAndServePodResources() {
 		return
 	}
-	server.ListenAndServePodResources(socket, kl.podManager, kl.containerManager, kl.containerManager, kl.containerManager)
+
```
@moshe010  thanks for the suggestion of splitting structs. I like this direction more. I think it would be impractical for you to wait for this PR to land, so if you also like this direction, feel free to cherry pick this commit (or rewrite it since it's very simple) in your PR

moshe010 replied:
Sure, will do :)

ffromani commented:
the PR is now revieable

cici37 commented:
/remove-sig api-machinery

ffromani commented:
/label api-review

API review requested for the kubelet configuration changes
the kubelet configuration changes are required to tune the rate limiting; support the rate limting was in turn identified as GA blocker:

https://github.com/kubernetes/enhancements/pull/3863
https://github.com/kubernetes/enhancements/pull/3791

SergeyKanzhelev commented:
/assign @swatisehgal

klueska commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -478,6 +478,11 @@ type KubeletConfiguration struct {
 	// If not specified the default value is ContainerRuntimeEndpoint
 	// +optional
 	ImageServiceEndpoint string
+
+	// PodresourcesLimits specifies the rate limits for the node-local podresources API endpoint
+	// +featureGate=KubeletPodResources
+	// +optional
+	PodresourcesLimits RateLimitConfiguration
```
`Podresources` or `PodResources`? We seem to have a mix of both, but we should be consistent. I think `PodResources` is probably more appropriate since it's two separate words.

ffromani replied:
Good point. I will update the PR.

liggitt replied:
agree with `PodResources` as the field name and `podResources` as the serialized json name (in v1beta1/types)

ffromani replied:
Thank you both. Fixed as suggested.

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -656,3 +661,16 @@ type MemorySwapConfiguration struct {
+// RateLimitConfiguration is used for setting the rate limits for the kubelet APIs exposed
+// locally on the node as gRPC over UNIX-domain sockets, like the podresources API
+type RateLimitConfiguration struct {
+	// MaxFrequency  defines the maximum allowed frequency of the calls represented
+	// as number of calls per second, cumulating all the calls.
+	// Omitting the value or setting to zero means no limit.
+	MaxFrequency float64
+	// MaxBurst defines the maximum numbers of calls that may happen at once, within
+	// the boundaries set by frequency.
+	// This value is ignored if the frequuency is unlimited.
+	MaxBurst int32
```
other rate limiter config uses `qps int32 / burst int32` ... is there a good reason to switch terms and types from that?

ffromani replied:
no good reason, changing.

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -656,3 +661,16 @@ type MemorySwapConfiguration struct {
+	// MaxFrequency  defines the maximum allowed frequency of the calls represented
+	// as number of calls per second, cumulating all the calls.
+	// Omitting the value or setting to zero means no limit.
+	MaxFrequency float64
```
what is the behavior when this limit is hit? does the last fetched result get replayed or does the caller get an error? if replayed, how is that indicated? if an error, how is the client expected to react, and what are the implications of clients being able to starve/disrupt each other?

liggitt replied:
how should admins tune this limit? we're generally trying to move away from fixed qps limits on other clients/servers since they're difficult to administer (in favor of things like priority-and-fairness in the API server)

liggitt commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -656,3 +661,16 @@ type MemorySwapConfiguration struct {
+	// MaxFrequency  defines the maximum allowed frequency of the calls represented
+	// as number of calls per second, cumulating all the calls.
+	// Omitting the value or setting to zero means no limit.
+	MaxFrequency float64
```
The kubelet serves lots of endpoints, many of which incur cost and make remote connections (pod exec/attach/logs, portforward, etc). Is there a reason we're targeting rate limits at a single one of those endpoints? Would be good to get an ack from kubelet leads that this config is coherent with the overall direction of kubelet serving.

ffromani replied:
we discussed this item (and https://github.com/kubernetes/kubernetes/pull/115852#discussion_r1126606012 and https://github.com/kubernetes/kubernetes/pull/115852#discussion_r1126605828) in sig-node meeting on March 7 and we don't have a conclusion just yet. We agreed to continue with the discussion here and/or on sig-node channel (cc @SergeyKanzhelev @dchen1107 )

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -820,6 +820,11 @@ type KubeletConfiguration struct {
+	// PodresourcesLimits specifies the rate limits for the node-local podresources API endpoint
+	// +featureGate=KubeletPodResources
+	// +optional
+	PodresourcesLimits RateLimitConfiguration `json:"podresourcesLimits,omitempty"`
```
this needs validating in ValidateKubeletConfiguration:
* must not be set if the podresources feature is disabled
* qps / burst may not be negative

ffromani replied:
thanks for pointing this out. Fixed.

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -1005,3 +1010,16 @@ type ExecEnvVar struct {
+	MaxFrequency float64 `json:"maxFrequency,omitempty"`
+	// MaxBurst defines the maximum numbers of calls that may happen at once, within
+	// the boundaries set by frequency.
+	// This value is ignored if the frequuency is unlimited.
```
need to define default behavior if unset/zero

```suggestion
	// This value is ignored if the frequency is unlimited.
```

liggitt commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -1005,3 +1010,16 @@ type ExecEnvVar struct {
+	// as number of calls per second, cumulating all the calls.
+	// Setting the values to zero means no limit.
+	MaxFrequency float64 `json:"maxFrequency,omitempty"`
+	// MaxBurst defines the maximum numbers of calls that may happen at once, within
+	// the boundaries set by frequency.
```
"within the boundaries set by frequency" is confusing ... usually burst allows temporarily exceeding qps

ffromani replied:
got it. Will rephrase.

swatisehgal commented:
/triage accepted
/priority important-soon 
(populated priority as this is being targeted for 1.27 release)

SergeyKanzhelev commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -478,6 +478,11 @@ type KubeletConfiguration struct {
 	// +optional
 	ImageServiceEndpoint string
+
+	// PodresourcesLimits specifies the rate limits for the node-local podresources API endpoint
```
My take is that to address the concern from the KEP we can have unconfigureable very high threshold to graduate the feature. It will protect from DDoS attacks, but will not be very limiting to scenarios.

If we will need configuration later we will need to create a common approach for all endpoints. Some sort of fair clients throttling may need to be implemented at that point.

ffromani replied:
thanks @SergeyKanzhelev for chiming in. I implemented your suggestion reusing a subset of the code proposed here: https://github.com/kubernetes/kubernetes/pull/116459
I'll cose this PR if the sig prefer to move in the direction of the hardcoded defaults, evaluating a the common approach later.

valaparthvi commented:
Hi @ffromani, I am from the Bug Triage Team for K8s release v1.27. I see this PR is in actively worked up, so I assume it is on track for this release.

However, please note  that Code Freeze begins on March 15, after which only critical and release blocker PR/issues will be accepted. Please make sure this PR goes in by March 15.

dims commented on pkg/kubelet/apis/config/types.go:
```diff
@@ -656,3 +661,16 @@ type MemorySwapConfiguration struct {
+	// MaxBurst defines the maximum numbers of calls that may happen at once, within
+	// the boundaries set by frequency.
+	// This value is ignored if the frequuency is unlimited.
+	MaxBurst int32
```
Looks like MaxBurst should be > 0 if frequency is NOT unlimited, looking at:
https://pkg.go.dev/golang.org/x/time/rate#:~:text=A%20zero%20Burst%20allows%20no%20events%2C%20unless%20limit%20%3D%3D%20Inf.

ffromani replied:
thanks for looking into this, will check and update

dims commented on pkg/kubelet/server/server.go:
```diff
@@ -209,11 +212,27 @@ func ListenAndServeKubeletReadOnlyServer(
+	maxCallsBurst := 0
+	if ratelimit.MaxFrequency > 0 {
+		maxCallsLimit = gotimerate.Limit(ratelimit.MaxFrequency)
+		maxCallsBurst = int(ratelimit.MaxBurst)
```
I think we should add a sanity check on MaxBurst (to ensure that MaxBurst is > 0) given my earlier comment - https://github.com/kubernetes/kubernetes/pull/115852/files#r1132523091

ffromani replied:
Good point, will add.

dims commented on test/e2e_node/podresources_test.go:
```diff
@@ -861,7 +863,59 @@ var _ = SIGDescribe("POD Resources [Serial] [Feature:PodResources][NodeFeature:P
+
+	ginkgo.Context("with the rate limit values configured", func() {
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			// we intentionally set very low limits to be reasonnably sure to hit them during the tests
```
`reasonnably` spelling

ffromani replied:
thanks, will be fixed in the next upload

dims commented on test/e2e_node/podresources_test.go:
```diff
@@ -861,7 +863,59 @@ var _ = SIGDescribe("POD Resources [Serial] [Feature:PodResources][NodeFeature:P
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			// we intentionally set very low limits to be reasonnably sure to hit them during the tests
+			initialConfig.PodresourcesLimits = kubeletconfig.RateLimitConfiguration{
+				MaxFrequency: 0.1,
```
can MaxFrequency < MaxBurst ?

ffromani replied:
Good point, I need to check.

ffromani commented on staging/src/k8s.io/kubelet/config/v1beta1/types.go:
```diff
@@ -1005,3 +1010,16 @@ type ExecEnvVar struct {
+	MaxFrequency float64 `json:"maxFrequency,omitempty"`
+	// MaxBurst defines the maximum numbers of calls that may happen at once, within
+	// the boundaries set by frequency.
+	// This value is ignored if the frequuency is unlimited.
```
thanks, will be fixed in the next upload

ffromani commented:
After discussion in sig-node, we don't have yet consensus about the very valid points emerged during the API review. Providing a generic rate-limiting covering all the endpoints, not just this one, is a good point which will need more research and work. For the much limited scope of the podresources API graduation we agreed to move into a different direction: https://github.com/kubernetes/kubernetes/pull/116459 . I'll keep this PR open for a little more until the final call is made (which will be happen very soon as the code freeze is looming).

ffromani commented:
#116459 was approved, we are going that direction.
Thanks @liggitt for the points raised in the API review. Implementing a general ratelimiting for the kubelet endpoint remains an important point, but indeed a much larger endeavour. Will require more discussion and, I believe, its own KEP, which should start addressing all the points raised in the review.
